### ISSUE TITLE
Foreign exchange instruments

### ISSUE THREAD

**Issue by jvirtanen** [paritytrading/parity#40]
The data type for prices in PMD, PMR and POE is a 32-bit unsigned integer with a six-digit integral part and a four-digit fractional part. This accuracy is not enough for foreign exchange instruments, including Bitcoin-related instruments.

To accommodate foreign exchange instruments, the number of digits in the fractional part should be made configurable. Furthermore, the data type for prices could be changed to a 64-bit unsigned integer.

**Comment by penberg** [comment]
@jvirtanen IIRC, the NYSE OpenBook format supports configurable fractional part. Perhaps worth taking a look?

**Comment by jvirtanen** [comment]
@penberg Thanks! I'll have a look at that.

**Comment by jvirtanen** [comment]
As a first step, the TAQ file format gets configurable fractional part for both prices and sizes in #74.

**Comment by jvirtanen** [comment]
[Genesis](https://github.com/jvirtanen/genesis) needs configurable fractional part for both prices and sizes.

Here is my current road map for implementing this feature:

- The trading system will operate on the minimum price and size increment.

- The network protocols will operate on the minimum price and size increment.

- The FIX gateway, terminal client, and stock ticker will get new configuration settings to specify the fractional part for both prices and sizes:
    ```
    instruments {
      AAPL {
        price-fragment-digits = 2
        size-fragment-digits  = 0
      },
      ETH-BTC {
        price-fragment-digits = 8
        size-fragment-digits  = 8
      },
    }
    ```

- The terminal client and stock ticker will get new configuration settings to specify the integer part for both prices and sizes for pretty printing:
    ```
    price-integer-digits = 8
    size-integer-digits  = 8
    ```

- The network protocols will start using 64-bit unsigned integers for price and size.

**Comment by jheusser** [comment]
Not sure this entirely belongs here, but worth considering how to represent null or max/min prices. For example in SBE you can define a null values as part of a composite type such as

```
<composite name="PRICENULL" description="Price NULL">
  <type name="mantissa" description="mantissa" presence="optional" nullValue="9223372036854775807" primitiveType="int64"/>
  <type name="exponent" description="exponent" presence="constant" primitiveType="int8">-7</type>
</composite>
```

**Comment by jvirtanen** [comment]
@jheusser That's a good point. Currently there's no concept of a null price, and tools, such as the stock ticker, rely on zero quantity to indicate missing data.

**Comment by jvirtanen** [comment]
See #91.

**Comment by jvirtanen** [comment]
Closed by #91.
